google/cloud/internal/rest_client.h

```diff
@@ -0,0 +1,120 @@
+class CurlImpl;
+
+// Provides methods corresponding to HTTP verbs to make requests to RESTful
+// services.
+class RestClient {
+ public:
+  virtual ~RestClient() = default;
+  virtual StatusOr<std::unique_ptr<RestResponse>> Delete(
+      RestRequest const& request) = 0;
+  virtual StatusOr<std::unique_ptr<RestResponse>> Get(
+      RestRequest const& request) = 0;
+  virtual StatusOr<std::unique_ptr<RestResponse>> Patch(
+      RestRequest const& request,
+      std::vector<absl::Span<char const>> const& payload) = 0;
+  virtual StatusOr<std::unique_ptr<RestResponse>> Post(
+      RestRequest const& request,
+      std::vector<absl::Span<char const>> const& payload) = 0;
+  virtual StatusOr<std::unique_ptr<RestResponse>> Put(
+      RestRequest const& request,
+      std::vector<absl::Span<char const>> const& payload) = 0;
+};
+
+// RestClient implementation using libcurl.
+class CurlRestClient : public RestClient {
+ public:
+  ~CurlRestClient() override = default;
+
+  CurlRestClient(CurlRestClient const&) = delete;
+  CurlRestClient(CurlRestClient&&) = default;
+  CurlRestClient& operator=(CurlRestClient const&) = delete;
+  CurlRestClient& operator=(CurlRestClient&&) = default;
+
+  StatusOr<std::unique_ptr<RestResponse>> Delete(
+      RestRequest const& request) override;
+  StatusOr<std::unique_ptr<RestResponse>> Get(
+      RestRequest const& request) override;
+  StatusOr<std::unique_ptr<RestResponse>> Patch(
+      RestRequest const& request,
+      std::vector<absl::Span<char const>> const& payload) override;
+  StatusOr<std::unique_ptr<RestResponse>> Post(
+      RestRequest const& request,
+      std::vector<absl::Span<char const>> const& payload) override;
+  StatusOr<std::unique_ptr<RestResponse>> Put(
+      RestRequest const& request,
+      std::vector<absl::Span<char const>> const& payload) override;
+
+ private:
+  friend class DefaultRestClient;
+  friend class PooledRestClient;
+
+  CurlRestClient(std::string endpoint_address,
+                 std::shared_ptr<CurlHandleFactory> factory, Options options);
+  StatusOr<std::unique_ptr<CurlImpl>> CreateCurlImpl(
+      RestRequest const& request);
+
+  std::string endpoint_address_;
+  std::shared_ptr<CurlHandleFactory> handle_factory_;
+  std::string x_goog_api_client_header_;
+  Options options_;
+};
+
+// Provides factory function to create a CurlRestClient that does not manage a
+// pool of connections.
+// This factory function is contained in a class in order to allow injection as
+// a template parameter for testing purposes.
+class DefaultRestClient {
+ public:
+  static std::unique_ptr<RestClient> GetRestClient(std::string endpoint_address,
+                                                   Options options);
+};
+
+// Provides factory function to create a CurlRestClient that manages a pool of
+// connections which are reused in order to minimize costs of setup and
+// teardown.
+// This factory function is contained in a class in order to allow injection as
```

Comment:
I cannot find any templates using `PooledRestClient` or `DefaultRestClient`.  In either case, seems like a `std::function<>` could serve that role in any tests?